New Sidebar Navigation

ramonski opened the pull request:
## Description of the issue/feature this PR addresses

This PR introduces a new SENAITE Sidebar Component, that can be configured in the new SENAITE Setup:

<img width="1436" height="482" alt="SCR-20251219-mcal" src="https://github.com/user-attachments/assets/e0e85764-22b9-4d6c-ba02-c9bc09de1c19" />

<img width="69" height="415" alt="SCR-20251219-mciy" src="https://github.com/user-attachments/assets/68a6a51c-1e85-4a40-89e8-98cb1fd190a7" />

### Features:

 - Toggle button for persistent state
 - Smooth CSS transitions
 - Keyboard navigation support (Ctrl/Cmd+B)
 - Collapsible navigation sections
 - Search functionality for navigation items
 - Resizable sidebar with drag handle


## Current behavior before PR

Server rendered sidebar used with limited functionalities

## Desired behavior after PR is merged

Client rendered ReactJS sidebar with enhanced functionalities

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

ramonski committed:
Initial commit

ramonski committed:
styles updated

ramonski committed:
Sidebar uses uid_catalog

ramonski committed:
Updated styles and JS

ramonski committed:
Move config to setup

ramonski committed:
Show top level items directly

ramonski committed:
Styles corrected

ramonski committed:
Handle UID catalog brains for path/url calculation

ramonski committed:
Refactoring

ramonski committed:
Fix sidebar search

ramonski committed:
Rebuild JS/CSS

ramonski committed:
Make sidebar resizable

ramonski committed:
Move sidebar into senaite.core namespace

ramonski committed:
Load more button styling

ramonski committed:
Load more functionality

ramonski committed:
Makde default depth 2

ramonski committed:
Handle show_more flag

ramonski committed:
Use specific catalog for child searches

ramonski committed:
Refactored to ReactJS components

ramonski committed:
Fixed build errors

ramonski committed:
Skip folders/portal_types in setup

ramonski committed:
Added tooltip

ramonski committed:
Changelog updated

ramonski committed:
Test fixed

ramonski committed:
production js

ramonski committed:
Mobile sidebar handling

ramonski committed:
Merge branch '2.x' into sidebar-navigation

ramonski committed:
Merge branch '2.x' into sidebar-navigation

ramonski committed:
Whitespace only

ramonski committed:
Merge branch 'sidebar-navigation' of https://github.com/senaite/senaite.core into sidebar-navigation

ramonski committed:
Merge branch '2.x' into sidebar-navigation

xispa commented on src/bika/lims/api/__init__.py:
```diff
@@ -817,8 +817,10 @@ def get_url(brain_or_object):
+    if is_brain(brain_or_object):
+        request = get_request()
+        path = get_path(brain_or_object)
+        return request.physicalPathToURL(path, relative=0)
```
Good point, we skip waking up the object if `getURL` is not a brain metadata!

xispa replied:
Good point, I see is better to rely on `get_path` and `physicalPathToURL` instead of waking up the object unnecessarily. Note to @xispa: remove existing calls to brain's `getURL` column (still present in listings like `AnalysesView` or `SamplesView`).

xispa commented on src/bika/lims/api/__init__.py:
```diff
@@ -952,7 +954,15 @@ def get_path(brain_or_object):
+        # Check if path is missing the portal path prefix
+        portal = get_portal()
+        portal_path = "/".join(portal.getPhysicalPath())
+        if not path.startswith(portal_path):
```
Truly not expensive, but given the fact that this is called for every single brain and we might do this a lot of times (e.g. in listings via `get_url`), do you think it would be worth to add a ram cache `api.get_portal_path` function?

ramonski replied:
In my opinion it is not worth it and I want to keep the API functions when possible without caching to avoid any side-effects.

xispa commented on src/senaite/core/config/setup.py:
```diff
@@ -0,0 +1,62 @@
+    "ATSimpleIntCriterion",
+    "ATSimpleStringCriterion",
+    "ATSortCriterion",
+    "TempFolder",
```
I would add `AnalysisRequest`, `Attachment` and `ARReport`. Just to be sure the system does not freeze when `sidebar_navigation_depth` is set to 3

ramonski replied:
I would like to leave that in the responsibility of the lab admin. Omitting them feels kind of limiting

xispa replied:
I understand this may feel somewhat limiting, but if a user clicks "Show more..." for a client with thousands of samples and results reports, the navigation bar may become unresponsive while displaying the "Loading navigation..." message. IMO, this is something we should prevent upfront, even if it is perceived as a _limitation_. Even if we don't apply this limitation, rendering the navbar with thousands of items makes the browser to suffer and the page is not displayed correctly. A part from that, we are already limiting to a depth of maximum 3.

ramonski replied:
I guess it would make sense to provide instead of "Sidebar displayed portal types" a field to blacklist display types and add there `AnalysisRequest` per default.

ramonski replied:
Switched the behavior in bc3d46f74 and added `AnalysisRequest` per default:

<img width="878" height="169" alt="SCR-20251229-sxxu" src="https://github.com/user-attachments/assets/576736f3-c050-43fa-a0e5-c21585652eb3" />

xispa commented on src/senaite/core/profiles/default/registry.xml:
```diff
@@ -87,11 +87,16 @@
     <value>
       <!-- Work Items -->
       <element>Samples</element>
+      <element>Sample</element>
```
We don't have the `Sample` portal_type yet, but `AnalysisRequest`

ramonski replied:
Note that we use here the `plone.app.vocabularies.ReallyUserFriendlyTypes` which resolves (if I remember correctly) our `AnalysisRequest` to `Sample`.
Haven't got it in front of me right now, but I think I did this by purpose...

xispa replied:
For what I've seen `ReallyUserFriendlyTypes` relies only on the registered portal types and uses a local `BAD_TYPES` variable to exclude some of them:

https://github.com/plone/plone.app.vocabularies/blob/52b90a8a5297f5d53a7e2ba2493bce2c8a51e76b/src/plone/app/vocabularies/types.py#L299-L303

A part from that, I think we don't use this vocabulary anywhere in SENAITE 🤔

ramonski replied:
I just see that we don't need that anymore 😅 Removed in 8b6871118

xispa commented on src/senaite/core/vocabularies/setup.py:
```diff
@@ -141,6 +144,57 @@ def __call__(self, context):
+    """
+
+    def __call__(self, context):
+        from bika.lims import api
```
This import is not necessary, is imported already at the top of the class

ramonski replied:
Removed in 137a8daa2

xispa commented on src/senaite/core/vocabularies/setup.py:
```diff
@@ -141,6 +144,57 @@ def __call__(self, context):
+            # Check if object is AT or DX
+            if not api.is_object(obj):
+                continue
+            if api.get_portal_type(obj) in SKIP_NAV_FOLDER_TYPES:
```
Not important at all However, I would not differentiate between `SKIP_NAV_FOLDER_TYPES` and `SKIP_NAV_PORTAL_TYPES` but instead, rely on `SKIP_NAV_PORTAL_TYPES` only.

ramonski replied:
You're right, thanks! Changed in 07d8d04fc

xispa reviewed (changes requested):
Excellent, just left two really small things. Will accept and merge afterwards

xispa committed:
Fix TypeError: Missing.Value is not JSON serializable

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.core.browser.viewlets.sidebar, line 179, in __call__
  Module json, line 244, in dumps
  Module json.encoder, line 207, in encode
  Module json.encoder, line 270, in iterencode
  Module json.encoder, line 184, in default
TypeError: Missing.Value is not JSON serializable
```

ramonski committed:
Removed inline import

ramonski committed:
Combined navigation types

ramonski committed:
Removed obsolete registry entry

ramonski committed:
Check brains for Missing.Value

ramonski committed:
Use skip types instead of display types

ramonski committed:
Fixed test

ramonski committed:
Skip Attachment type as well

ramonski committed:
Removed accessibility attributes

xispa reviewed (approved):
Excellent, thanks!
